test/correctness/callable_generator.cpp

```diff
@@ -0,0 +1,253 @@
+                in1(i, j) = i + j * 10;
+                in2(i, j) = i * 10 + j;
+            }
+        }
+
+        auto gen = TestGen1::create(context);
+        Callable c = gen->compile_to_callable();
+
+        Buffer<uint8_t> out1(10, 10);
+        check(c(in1, 42, 1.0f, out1));
+
+        Buffer<uint8_t> out2(10, 10);
+        check(c(in2, 22, 2.0f, out2));
+
+        Buffer<uint8_t> out3(10, 10);
+        check(c(in1, 12, 1.0f, out3));
+
+        Buffer<uint8_t> out4(10, 10);
+        check(c(in2, 16, 1.0f, out4));
+
+        for (int i = 0; i < 10; i++) {
+            for (int j = 0; j < 10; j++) {
+                assert(out1(i, j) == i + j * 10 + 42);
+                assert(out2(i, j) == i * 10 + j + 11);
+                assert(out3(i, j) == i + j * 10 + 12);
+                assert(out4(i, j) == i * 10 + j + 16);
+            }
+        }
+
+        // Test bounds inference
+        Buffer<uint8_t> in_bounds(nullptr, 1, 1);
+        Buffer<uint8_t> out_bounds(nullptr, 20, 20);
+
+        check(c(in_bounds, 42, 1.0f, out_bounds));
+
+        assert(in_bounds.defined());
+        assert(in_bounds.dim(0).extent() == 20);
+        assert(in_bounds.dim(1).extent() == 20);
+        assert(in1.dim(0).extent() == 10);
+        assert(in1.dim(1).extent() == 10);
+    }
+
+    // Override Halide's malloc and free (except under wasm),
+    // make sure that Callable freezes the values
+    if (t.arch != Target::WebAssembly) {
+        custom_malloc_called = false;
+        custom_free_called = false;
+
+        class TestGen2 : public Generator<TestGen2> {
+        public:
+            Output<Buffer<int, 1>> out_{"out"};
+
+            void generate() {
+                Var x("x");
+
+                Func f;
+                f(x) = x;
+
+                out_(x) = f(x);
+
+                f.compute_root();
+            }
+        };
+
+        JITHandlers my_jit_handlers;
+        my_jit_handlers.custom_malloc = my_malloc;
+        my_jit_handlers.custom_free = my_free;
+
+        auto gen = TestGen2::create(context);
+        Callable c = gen->compile_to_callable(&my_jit_handlers);
+
+        Buffer<int> im(100000);
+        check(c(im));
+
+        assert(custom_malloc_called);
+        assert(custom_free_called);
+    }
+
+    // Check that Param<void*> works with Callables
+    if (t.arch != Target::WebAssembly) {
+        class TestGen3 : public Generator<TestGen3> {
+        public:
+            GeneratorParam<bool> vectorize_{"vectorize", false};
+
+            Input<void *> handle_{"handle"};
+            Output<Buffer<uint64_t, 1>> out_{"out"};
+
+            void generate() {
+                Var x("x");
+
+                out_(x) = reinterpret<uint64_t>(handle_);
+                if (vectorize_) {
+                    out_.vectorize(x, 4);
+                }
+            }
+        };
+
+        auto gen_1 = TestGen3::create(context);
+        gen_1->set_generatorparam_values({{"vectorize", "false"}});
```

Comment:
This is a nit, but IMO it would good to change it to avoid the unnecessary string-typing.

Reply:
I'm gonna push back on the string-typing concern a bit:
- This uses the same syntax that build systems would use, which IMHO is a good thing
- Setting the generator param here via C++ would work (because we know the complete type), but it wouldn't work for the look-up-by-registered-name version because that returns an AbstractGenerator. (This syntax works with both).

Reply:
IMO This should serve as an example of how to get a callable out of a generator declared in scope. If you look up by registered string name it's reasonable to set the generator params by string, but if you create the generator by actual C++ name, you should also set the generator params by actual c++ name. There's no reason to make yourself vulnerable to typos that could have been caught at C++ compile time here.

I think the build system isn't that relevant because we're talking about jit usage of an in-scope generator, which isn't going to show up in the build system at all. Even if it did, in the build system it would look like `vectorize=false`, not `{{"vectorize", "false"}}`. It's debatable whether or not `vectorize.set(false)` is closer to one or the other.

Reply:
Fine, will change